cpp/src/ast/transform.cu

```diff
@@ -0,0 +1,168 @@
+  auto const start_idx             = cudf::size_type(threadIdx.x + blockIdx.x * blockDim.x);
+  auto const stride                = cudf::size_type(blockDim.x * gridDim.x);
+  auto const num_rows              = table.num_rows();
+  auto const evaluator =
+    cudf::ast::detail::row_evaluator(table, literals, thread_intermediate_storage, &output_column);
+
+  for (cudf::size_type row_index = start_idx; row_index < num_rows; row_index += stride) {
+    evaluate_row_expression(
+      evaluator, data_references, operators, operator_source_indices, num_operators, row_index);
+  }
+}
+
+std::unique_ptr<column> compute_column(table_view const table,
+                                       expression const& expr,
+                                       cudaStream_t stream,
+                                       rmm::mr::device_memory_resource* mr)
+{
+  // Linearize the AST
+  auto const expr_linearizer         = linearizer(expr, table);
+  auto const data_references         = expr_linearizer.get_data_references();
+  auto const literals                = expr_linearizer.get_literals();
+  auto const operators               = expr_linearizer.get_operators();
+  auto const num_operators           = cudf::size_type(operators.size());
+  auto const operator_source_indices = expr_linearizer.get_operator_source_indices();
+  auto const expr_data_type          = expr_linearizer.get_root_data_type();
+
+  // Create ast_plan and device buffer
+  auto plan = ast_plan();
+  plan.add_to_plan(data_references);
+  plan.add_to_plan(literals);
+  plan.add_to_plan(operators);
+  plan.add_to_plan(operator_source_indices);
+  auto const host_data_buffer = plan.get_host_data_buffer();
+  auto const buffer_offsets   = plan.get_offsets();
+  auto const buffer_size      = host_data_buffer.second;
+  auto device_data_buffer     = rmm::device_buffer(buffer_size, stream, mr);
+  CUDA_TRY(cudaMemcpyAsync(device_data_buffer.data(),
+                           host_data_buffer.first.get(),
+                           buffer_size,
+                           cudaMemcpyHostToDevice,
+                           stream));
+  // To reduce overhead, we don't call a stream sync here.
+  // The stream is synced later when the table_device_view is created.
+
+  // Create device pointers to components of plan
+  auto const device_data_buffer_ptr = static_cast<const char*>(device_data_buffer.data());
+  auto const device_data_references = reinterpret_cast<const detail::device_data_reference*>(
+    device_data_buffer_ptr + buffer_offsets.at(0));
+  auto const device_literals =
+    reinterpret_cast<const cudf::detail::fixed_width_scalar_device_view_base*>(
+      device_data_buffer_ptr + buffer_offsets.at(1));
+  auto const device_operators =
+    reinterpret_cast<const ast_operator*>(device_data_buffer_ptr + buffer_offsets.at(2));
+  auto const device_operator_source_indices =
+    reinterpret_cast<const cudf::size_type*>(device_data_buffer_ptr + buffer_offsets.at(3));
```

Comment:
Is this still TODO?

Reply:
tl;dr This is not done yet. This would be a good "cleanup" task but does not impact functionality.

There was a problem when I moved this logic to the constructor, also described in [this comment](https://github.com/rapidsai/cudf/pull/5494/files#r475286514). In commit d8479b4, I moved this logic to the `ast_plan` constructor, but I did not consider the lifetime of the host-side buffer held by the `ast_plan`, which meant the data was not async-copied to the device before going out of scope. I was being careful to avoid overhead here and left out the stream sync (mentioned in the code comment above). Adding an explicit stream sync after constructing the device buffer in that commit (which triggers an async copy from the host buffer) would allow this to be done in the constructor without worrying about scope: https://github.com/rapidsai/cudf/blob/d8479b4b4f5e11dd4573f7af54654f798b613e8b/cpp/include/cudf/ast/detail/transform.cuh#L391-L392

The additional overhead of a stream sync is measurable for small data sizes like O(100k) rows, where the host-device copy of the `ast_plan` and synchronization takes roughly as much time as the kernel execution itself.